frontend/src/firebase/actions.ts

```diff
@@ -152,21 +139,51 @@ export const removeAllForks = (taskId: string): void => {
   batch.commit().then(ignore);
 };
 
+type EditType = 'EDITING_MASTER_TEMPLATE' | 'EDITING_ONE_TIME_TASK' | 'FORKING_MASTER_TEMPLATE';
+
 export const editTaskWithDiff = (
   taskId: string,
-  forMasterTemplate: boolean,
+  editType: EditType,
   { mainTaskEdits, subTaskCreations, subTaskEdits, subTaskDeletions }: Diff,
 ): void => {
   const batch = db().batch();
-  if (forMasterTemplate) {
+  let updateTaskId = taskId;
+  if (editType === 'FORKING_MASTER_TEMPLATE') {
+    const { tasks, subTasks } = store.getState();
+    const repeatingTask = tasks.get(taskId) as RepeatingTask;
+    (async () => {
+      const createdSubTasks = repeatingTask.children.map((child) => {
+        const subtask = subTasks.get(child);
+        const firebaseSubTask: FirestoreSubTask = mergeWithOwner(subtask);
+        const subtaskDoc = subTasksCollection().doc();
+        batch.set(subtaskDoc, firebaseSubTask);
+        return { ...subtask, id: subtaskDoc.id };
+      });
+      await batch.commit();
+      const subtaskIds = createdSubTasks.map(s => s.id);
+      const taskWithChildren = { ...repeatingTask, children: subtaskIds };
+      const firestoreTask: FirestoreTask = await createFirestoreObject('tasks', taskWithChildren);
+      const addedDoc = await tasksCollection().add(firestoreTask);
+      const newForkMetaData = {
+        forkId: addedDoc.id,
+        replaceDate: mainTaskEdits.date !== undefined ? mainTaskEdits.date : repeatingTask.date,
+      };
+      batch.update(tasksCollection().doc(taskId), {
+        forks: [...repeatingTask.forks, newForkMetaData],
+      });
+      updateTaskId = addedDoc.id;
+    })();
+  }
+
+  if (editType === 'EDITING_MASTER_TEMPLATE') {
     removeAllForks(taskId);
```

Comment:
You will enter a race condition below. If you look at the implementation of `removeAllForks`, it is not a blocking function. I think the easiest way is to change `removeAllForks` into an async function, and then `await` for `await removeAllForks(taskId)` (although you might need to make the entire block starting here to be `async`).